types/underscore/index.d.ts

```diff
@@ -234,76 +239,139 @@ declare module _ {
+         * @returns The reduced result.
+         **/
+        reduce<V extends Collection<any>>(
+            collection: V,
+            iteratee: MemoCollectionIterator<TypeOfCollection<V>, TypeOfCollection<V>, V>,
+            memo?: undefined,
+            context?: any
+        ): TypeOfCollection<V>;
+
+        /**
+         * Also known as inject and foldl, reduce boils down a collection of values into a
+         * single value. Memo is the initial state of the reduction, and each successive
+         * step of it should be returned by iteratee. The iteratee is passed four arguments:
+         * the memo, then the value and index (or key) of the iteration, and finally a reference
+         * to the entire collection.
+         *
+         * If no memo is passed to the initial invocation of reduce, the iteratee is not invoked
+         * on the first element of the collection. The first element is instead passed as the memo
+         * in the invocation of the iteratee on the next element in the collection.
+         * @param collection Reduces the elements of this collection.
+         * @param iteratee Reduce iteratee function for each element in `collection`.
+         * @param memo Initial reduce state or undefined to use the first collection item as initial state.
+         * @param context `this` object in `iteratee`, optional.
+         * @returns The reduced result.
+         **/
+        reduce<V extends List<any>, TResult>(
+            collection: V,
+            iteratee: MemoIterator<TypeOfCollection<V>, TResult, V>,
+            memo: TResult,
+            context?: any
+        ): TResult;
+
+        /**
+         * Also known as inject and foldl, reduce boils down a collection of values into a
+         * single value. Memo is the initial state of the reduction, and each successive
+         * step of it should be returned by iteratee. The iteratee is passed four arguments:
+         * the memo, then the value and index (or key) of the iteration, and finally a reference
+         * to the entire collection.
+         *
+         * If no memo is passed to the initial invocation of reduce, the iteratee is not invoked
+         * on the first element of the collection. The first element is instead passed as the memo
+         * in the invocation of the iteratee on the next element in the collection.
+         * @param collection Reduces the elements of this collection.
+         * @param iteratee Reduce iteratee function for each element in `collection`.
+         * @param memo Initial reduce state or undefined to use the first collection item as initial state.
+         * @param context `this` object in `iteratee`, optional.
+         * @returns The reduced result.
+         **/
+        reduce<V extends Dictionary<any>, TResult>(
+            collection: V,
+            iteratee: MemoObjectIterator<TypeOfCollection<V>, TResult, V>,
+            memo: TResult,
+            context?: any
+        ): TResult;
```

Comment:
One of them is for lists and one of them is for dictionaries I think, though I may have messed that up. The reason I'm doing that rather than using collection constraints and iterators is that TS is being odd with its type inference of `TResult` in that case (see the difference in inferred `TResult` types in the comments at the bottom of [this simplified example](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/45893#discussion_r449856859)), which is something I might need to file an issue with the TS folks to figure out unfortunately.